### PR TITLE
Java Net Authenticator

### PR DESCRIPTION
**Pull request by yschimke**
Refactor the okhttp-urlconnection implementation.  Available as 

```
client.newBuilder()
        .proxyAuthenticator(okhttp3.Authenticator.JAVA_NET_AUTHENTICATOR)
        .build()
```

### CODE REVIEW

File: okhttp/src/main/kotlin/okhttp3/internal/authenticator/JavaNetAuthenticator.kt [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+      }
+    }
+
+    return null // No challenges were satisfied!
+  }
+
+  @Throws(IOException::class)
+  private fun Proxy.connectToInetAddress(url: HttpUrl, dns: Dns): InetAddress {
+    return when (type()) {
+      Proxy.Type.DIRECT -> dns.lookup(url.host).first()
```
Comment (yschimke) [inline okhttp/src/main/kotlin/okhttp3/internal/authenticator/JavaNetAuthenticator.kt:90]: Changed from existing java impl.

Comment (swankjesse) [reply okhttp/src/main/kotlin/okhttp3/internal/authenticator/JavaNetAuthenticator.kt:90]: Nice improvement

File: okhttp/src/main/kotlin/okhttp3/Authenticator.kt [hunk 2]
```diff
@@ -114,5 +115,9 @@ interface Authenticator {
     private class AuthenticatorNone : Authenticator {
       override fun authenticate(route: Route?, response: Response): Request? = null
     }
+
+    /** An authenticator that knows no credentials and makes no attempt to authenticate. */
```
Comment (swankjesse) [inline okhttp/src/main/kotlin/okhttp3/Authenticator.kt:119]: Copy pasta problem

This has credentials; they’re installed in `java.net.Authenticator` !

File: okhttp/src/test/java/okhttp3/internal/authenticator/JavaNetAuthenticatorTest.kt [hunk 3]
```diff
@@ -0,0 +1,74 @@
+package okhttp3.internal.authenticator
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/internal/authenticator/JavaNetAuthenticatorTest.kt:1]: supernit: copyright

File: okhttp/src/test/java/okhttp3/internal/authenticator/JavaNetAuthenticatorTest.kt [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        .message("Unauthorized")
+        .build()
+    val authRequest = authenticator.authenticate(route, response)
+
+    assertNotNull(authRequest)
+    assertEquals(
+        "Basic ${RecordingAuthenticator.BASE_64_CREDENTIALS}", authRequest!!.header("Authorization")
+    )
+  }
+}
```
Comment (monkey-mas) [inline okhttp/src/test/java/okhttp3/internal/authenticator/JavaNetAuthenticatorTest.kt:74]: nit: maybe we can add another test for non-Basic auth request? I might be wrong though.
```kotlin
  @Test
  fun noSupportForNonBasicAuth() {
    val request = Request.Builder()
        .url("https://server/robots.txt")
        .build()

    val response = Response.Builder()
        .request(request)
        .code(401)
        .header("WWW-Authenticate", "UnsupportedScheme realm=\"User Visible Realm\"")
        .protocol(HTTP_2)
        .message("Unauthorized")
        .build()

    val authRequest = authenticator.authenticate(null, response)
    assertNull(authRequest)
  }
```

File: okhttp/src/test/java/okhttp3/internal/authenticator/JavaNetAuthenticatorTest.kt [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    val response = Response.Builder()
+        .request(request)
+        .code(401)
+        .header("WWW-Authenticate", "Basic realm=\"User Visible Realm\"")
+        .protocol(HTTP_2)
+        .message("Unauthorized")
+        .build()
+    val authRequest = authenticator.authenticate(route, response)
+
+    assertNotNull(authRequest)
```
Comment (monkey-mas) [inline okhttp/src/test/java/okhttp3/internal/authenticator/JavaNetAuthenticatorTest.kt:69]: supernit: given that `authRequest!!.header("Authorization")` throws NPE when `authRequest` is null, we could drop this non-null assertion.

File: okhttp/src/test/java/okhttp3/internal/authenticator/JavaNetAuthenticatorTest.kt (hunk 3 above)
Comment (yschimke) [reply okhttp/src/test/java/okhttp3/internal/authenticator/JavaNetAuthenticatorTest.kt:1]: We should automate this in spotless :)

Would save you a lot of typing on my reviews :)

**Pull request merged**